### PR TITLE
Remove UNKNOWN_NULL_COUNT

### LINKED ISSUE #11968: [FEA] Require null counts on column construction
**Is your feature request related to a problem? Please describe.**
Currently the null count may be provided when constructing a column, or it may be left unknown to be computed lazily when it is needed. This approach has performance benefits in some cases, but it is problematic for exposing a stream-ordered API because null counts may be used in contexts where a stream is not available and requesting the null count instead triggers a kernel launch on cudf's default stream. Additionally, the current approach sometimes leads to calling code not providing a null count even though it is known, resulting in an unnecessary extra computation.

**Describe the solution you'd like**
We should change the column constructor to require the null count on construction. This change will guarantee that requesting the null count does not result in a kernel launch, which removes the most problematic known blocker for implementing a stream-ordered API in libcudf. We will also need to make the same change to `column_view`. When constructed from a `column`, the `column_view`'s null count will now be trivially available provided, but when constructed from an external data pointer it will again be the caller's responsibility to provide the null count.

To support these new requirements, we should expose public utility functions for computing the null count given a null mask.

**Describe alternatives you've considered**
We could instead have the column/column_view constructors accept the null count _optionally_ and then compute the null count if one is not provided. However, that would also necessitate accepting a stream in the constructor that _may_ be used. The resulting API is more confusing and requires more logic, whereas the calling code already has all the information available to make the best decision here.

**Additional context**
There are certain cases that will be made slower by this requirement. Two cases in particular come to mind that we will need to be cognizant of and attempt to mitigate the costs:
1. Places where null counts are not needed. In such cases, we will now be requiring a computation that could previously be omitted entirely.
2. The creation of slices. Slices will also now need to always know their null counts, which could add significant costs to a number of APIs relying on slices. These functions may need to be rewritten to more intelligently precompute the null counts, or be otherwise modified to reduce and/or hide the costs of this extra work.

### PR DESCRIPTION
**Pull request by vyasr**
## Description

This is the final PR for removing `UNKNOWN_NULL_COUNT` and the implicit kernel launch in the `null_count` methods of `column` and `column_view`. 

Depends on #13355 and #13341.

Closes #11968

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by vyasr** [comment]
There are a couple of Java failures left that I can reproduce locally. Will attempt to debug them myself, but may need to pull in some advice.

File: cpp/tests/groupby/var_tests.cpp [hunk 1]
```diff
@@ -167,3 +167,34 @@ TYPED_TEST(groupby_var_test, dictionary)
...
+  auto result1 = gb_obj.aggregate(requests);
+
+  // This agg forces a sort groupby.
+  auto agg2 = cudf::make_quantile_aggregation<cudf::groupby_aggregation>({0.25});
+  requests[0].aggregations.push_back(std::move(agg2));
+
+  auto result2 = gb_obj.aggregate(requests);
+
+  EXPECT_EQ(result1.second[0].results[0]->null_count(), result2.second[0].results[0]->null_count());
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(*result1.second[0].results[0], *result2.second[0].results[0]);
```
Comment (davidwendt) [inline cpp/tests/groupby/var_tests.cpp:199]: The `CUDF_TEST_EXPECT_COLUMNS_EQUAL` includes comparing the null counts so I don't think you need the `EXPECT_EQ`.

File: cpp/include/cudf/column/column_view.hpp [hunk 2]
```diff
@@ -511,10 +498,6 @@ class mutable_column_view : public detail::column_view_base {
    * @brief Construct a `mutable_column_view` from pointers to device memory for
    *the elements and bitmask of the column.
```
Comment (davidwendt) [inline cpp/include/cudf/column/column_view.hpp:499]:
```suggestion
   * the elements and bitmask of the column.
```

File: cpp/src/groupby/sort/group_std.cu [hunk 3]
```diff
@@ -20,12 +20,14 @@
 #include <cudf/column/column_factories.hpp>
 #include <cudf/column/column_view.hpp>
 #include <cudf/detail/aggregation/aggregation.hpp>
+#include <cudf/detail/null_mask.hpp>
```
Comment (davidwendt) [inline cpp/src/groupby/sort/group_std.cu:23]:
```suggestion
```
Probably not needed?

File: cpp/tests/groupby/var_tests.cpp (hunk 1 above)
Comment (vyasr) [reply cpp/tests/groupby/var_tests.cpp:199]: Ah yes this was a copy-paste from an earlier iteration.

File: cpp/src/groupby/sort/group_std.cu (hunk 3 above)
Comment (vyasr) [reply cpp/src/groupby/sort/group_std.cu:23]: Yup this was for `count_nulls` before I switched to the `atomicAdd` approach.

**Review by mythrocks — APPROVED** [review]
I'm +1 on the JNI changes. 

Note that `spark-rapids` does not depend on this code. `row_conversion.cu` will be removed as part of https://github.com/rapidsai/cudf/issues/13374. The splash damage here should be limited.

File: cpp/src/column/column.cu [hunk 4]
```diff
@@ -117,37 +117,16 @@ mutable_column_view column::mutable_view()
...
-size_type column::null_count() const
-{
-  CUDF_FUNC_RANGE();
-  if (_null_count <= cudf::UNKNOWN_NULL_COUNT) {
-    _null_count = cudf::detail::null_count(
-      static_cast<bitmask_type const*>(_null_mask.data()), 0, size(), cudf::get_default_stream());
-  }
-  return _null_count;
-}
+size_type column::null_count() const { return _null_count; }
```
Comment (ttnghia) [inline cpp/src/column/column.cu:129]: This may be moved to the header for better inline.

File: cpp/src/column/column_view.cpp [hunk 5]
```diff
@@ -63,15 +63,7 @@ column_view_base::column_view_base(data_type type,
...
-// If null count is known, returns it. Else, compute and return it
-size_type column_view_base::null_count() const
-{
-  if (_null_count <= cudf::UNKNOWN_NULL_COUNT) {
-    _null_count = cudf::detail::null_count(
-      null_mask(), offset(), offset() + size(), cudf::get_default_stream());
-  }
-  return _null_count;
-}
+size_type column_view_base::null_count() const { return _null_count; }
```
Comment (ttnghia) [inline cpp/src/column/column_view.cpp:66]: This may be moved to the header for better inline.

File: cpp/src/column/column.cu (hunk 4 above)
Comment (vyasr) [reply cpp/src/column/column.cu]: Done.

**Review by ttnghia — APPROVED** [review]

**Review by karthikeyann — APPROVED** [review]

**Pull request merged**